trio_run_in_process/run_in_process.py

```diff
@@ -223,3 +224,65 @@ async def open_worker_process() -> AsyncIterator[WorkerProcessAPI]:
                 yield worker
     finally:
         worker._dead = True
+
+
+class WorkerPool(WorkerPoolAPI):
+    def __init__(self, max_workers: int) -> None:
+        self._max_workers = max_workers
+        self._num_workers = 0
+        self._send_channel, self._receive_channel = trio.open_memory_channel[
+            WorkerProcessAPI
+        ](max_workers)
+        self._exit_stack = AsyncExitStack()
+        self._is_open = False
+
+    @asynccontextmanager
+    async def _reserve_worker(self) -> AsyncIterator[WorkerProcessAPI]:
+        try:
+            worker = self._receive_channel.receive_nowait()
+        except trio.WouldBlock:
+            if self._num_workers < self._max_workers:
+                self._num_workers += 1
+                logger.debug("Created new worker")
+                worker = await self._exit_stack.enter_async_context(
+                    open_worker_process()
+                )
+            else:
+                logger.debug("Waiting for a busy worker to become free")
+                worker = await self._receive_channel.receive()
+
+        try:
+            yield worker
+        finally:
+            logger.debug("Worker became free, making it available to other tasks")
+            self._send_channel.send_nowait(worker)
+
+    async def run(self, async_fn: Callable[..., TReturn], *args: Any) -> TReturn:
+        if not self._is_open:
+            raise Exception("WorkerPool is not open")
+
+        async with self._reserve_worker() as worker:
+            logger.debug("Got free worker, running %s(%s)", async_fn, args)
+            return await worker.run(async_fn, *args)
+
+    @asynccontextmanager
+    async def _open(self) -> AsyncIterator[WorkerPoolAPI]:
+        self._is_open = True
+        try:
+            yield self
+        finally:
+            self._is_open = False
+            await self._exit_stack.aclose()
```

Comment:
I think this function is more intuitive if done as:

```python
async def _open(self):
    async with self._exit_stack:
        self._is_open = True
        try:
            yield self
        finally:
            self._is_open = False
```

Reply:
Indeed!